Added support for quadrant chart

amsubhash opened the pull request:
## :bookmark_tabs: Summary

This PR contains the implementation of the new chart type `quadrant-chart` 

Resolves #4282

## :straight_ruler: Design Decisions

I have used a builder class `quadrantBuilder.ts` which is responsible for the final data creation and space calculation.

data:
```
    %%{init: {"quadrantChart": {"quadrantPadding": 10}, "theme": "forest", "themeVariables": {"quadrant1TextFill": "blue"}} }%%
    quadrantChart
      x-axis Urgent --> Not Urgent
      y-axis Not Important --> important
      quadrant-1 Plan
      quadrant-2 Do
      quadrant-3 Deligate
      quadrant-4 Delete
```
Diagram:
<img width="511" alt="image" src="https://github.com/mermaid-js/mermaid/assets/7448177/86204ec2-0638-41ff-9232-a7d78cd4c06b">

data:
```
    %%{init: {"quadrantChart": {"chartWidth": 600, "chartHeight": 600} } }%%
    quadrantChart
      title Analytics and Business Intelligence Platforms
      x-axis "Completeness of Vision ❤" -->
      y-axis Ability to Execute
      quadrant-1 Leaders
      quadrant-2 Challengers
      quadrant-3 Niche
      quadrant-4 Visionaries
      Microsoft: [0.75, 0.75]
      Salesforce: [0.55, 0.60]
      IBM: [0.51, 0.40]
      Incorta: [0.20, 0.30]
```
Diagram:
<img width="617" alt="image" src="https://github.com/mermaid-js/mermaid/assets/7448177/a93a77e4-fc27-4842-aed6-f5089667e049">

### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
- [x] :computer: have added unit/e2e tests (if appropriate)
- [x] :notebook: have added documentation (if appropriate)
- [x] :bookmark: targeted `develop` branch

sidharthv96 commented on packages/mermaid/src/diagrams/quadrant-chart/parser/quadrant.jison:
```diff
@@ -0,0 +1,248 @@
+[\u1FD6-\u1FDB\u1FE0-\u1FEC\u1FF2-\u1FF4\u1FF6-\u1FFC\u2071\u207F]|
+[\u2090-\u209C\u2102\u2107\u210A-\u2113\u2115\u2119-\u211D\u2124\u2126\u2128]|
+[\u212A-\u212D\u212F-\u2139\u213C-\u213F\u2145-\u2149\u214E\u2183\u2184]|
+[\u2C00-\u2C2E\u2C30-\u2C5E\u2C60-\u2CE4\u2CEB-\u2CEE\u2CF2\u2CF3]|
+[\u2D00-\u2D25\u2D27\u2D2D\u2D30-\u2D67\u2D6F\u2D80-\u2D96\u2DA0-\u2DA6]|
+[\u2DA8-\u2DAE\u2DB0-\u2DB6\u2DB8-\u2DBE\u2DC0-\u2DC6\u2DC8-\u2DCE]|
+[\u2DD0-\u2DD6\u2DD8-\u2DDE\u2E2F\u3005\u3006\u3031-\u3035\u303B\u303C]|
+[\u3041-\u3096\u309D-\u309F\u30A1-\u30FA\u30FC-\u30FF\u3105-\u312D]|
+[\u3131-\u318E\u31A0-\u31BA\u31F0-\u31FF\u3400-\u4DB5\u4E00-\u9FCC]|
+[\uA000-\uA48C\uA4D0-\uA4FD\uA500-\uA60C\uA610-\uA61F\uA62A\uA62B]|
+[\uA640-\uA66E\uA67F-\uA697\uA6A0-\uA6E5\uA717-\uA71F\uA722-\uA788]|
+[\uA78B-\uA78E\uA790-\uA793\uA7A0-\uA7AA\uA7F8-\uA801\uA803-\uA805]|
+[\uA807-\uA80A\uA80C-\uA822\uA840-\uA873\uA882-\uA8B3\uA8F2-\uA8F7\uA8FB]|
+[\uA90A-\uA925\uA930-\uA946\uA960-\uA97C\uA984-\uA9B2\uA9CF\uAA00-\uAA28]|
+[\uAA40-\uAA42\uAA44-\uAA4B\uAA60-\uAA76\uAA7A\uAA80-\uAAAF\uAAB1\uAAB5]|
+[\uAAB6\uAAB9-\uAABD\uAAC0\uAAC2\uAADB-\uAADD\uAAE0-\uAAEA\uAAF2-\uAAF4]|
+[\uAB01-\uAB06\uAB09-\uAB0E\uAB11-\uAB16\uAB20-\uAB26\uAB28-\uAB2E]|
+[\uABC0-\uABE2\uAC00-\uD7A3\uD7B0-\uD7C6\uD7CB-\uD7FB\uF900-\uFA6D]|
+[\uFA70-\uFAD9\uFB00-\uFB06\uFB13-\uFB17\uFB1D\uFB1F-\uFB28\uFB2A-\uFB36]|
+[\uFB38-\uFB3C\uFB3E\uFB40\uFB41\uFB43\uFB44\uFB46-\uFBB1\uFBD3-\uFD3D]|
+[\uFD50-\uFD8F\uFD92-\uFDC7\uFDF0-\uFDFB\uFE70-\uFE74\uFE76-\uFEFC]|
+[\uFF21-\uFF3A\uFF41-\uFF5A\uFF66-\uFFBE\uFFC2-\uFFC7\uFFCA-\uFFCF]|
+[\uFFD2-\uFFD7\uFFDA-\uFFDC]
+                                                          return 'UNICODE_TEXT';
```
We don't need this unicode block. Unicode symbols can be added inside `""`.

sidharthv96 commented on packages/mermaid/src/diagrams/quadrant-chart/quadrantDb.js:
Can you use typescript, with proper types?

sidharthv96 commented on packages/mermaid/src/diagrams/quadrant-chart/quadrantRenderer.js:
Typescript. We're slowly rewriting everything to TS, so new files should be in TS.

amsubhash replied:
converted

sidharthv96 commented on packages/mermaid/src/diagrams/quadrant-chart/styles.js:
You don't need a dummy file. You can set it in DiagramDefinition.
```ts
styles: () => ''
```

sidharthv96 reviewed (changes requested):
Looks like an excellent start to me!
Docs and Tests need to be added, along with proper typings.

Also, please add some screenshots in the PR.

sidharthv96 commented on packages/mermaid/src/diagrams/quadrant-chart/quadrantDb.ts:
```diff
@@ -0,0 +1,133 @@
+  quadrantBuilder.yAxisBottomText = textSanitizer(textObj.text);
+}
+
+function addPoints(textObj: LexTextObj, x: number, y: number) {
```
```suggestion
function addPoint(textObj: LexTextObj, x: number, y: number) {
```

amsubhash replied:
changed

sidharthv96 commented on packages/mermaid/src/diagrams/quadrant-chart/quadrantBuilder.ts:
```diff
@@ -0,0 +1,385 @@
+  private _yAxisTopText = '';
+  private _totalHeight = 500;
+  private _totalWidth = 500;
+  public quadrantPadding = 5;
+  public xAxisLabelPadding = 5;
+  public yAxisLabelPadding = 5;
+  public xAxisLabelFontSize = 16;
+  public yAxisLabelFontSize = 16;
+  public quadrantLabelFontSize = 16;
+  public quadrantTextTopPadding = 5;
+  public pointTextPadding = 5;
+  public pointLabelFontSize = 12;
+  public pointRadius = 5;
+  public points: QuadrantPointInputType[] = [];
+  public xAxisPosition = 'top';
+  public yAxisPosition = 'left';
+  public quadrant1Fill = '#8bc2f3';
+  public quadrant2Fill = '#faebd7';
+  public quadrant3Fill = '#00ffff';
+  public quadrant4Fill = '#f0ffff';
+  public quadrant1TextFill = '#93690e';
+  public quadrant2TextFill = '#8644ff';
+  public quadrant3TextFill = '#e3004d';
+  public quadrant4TextFill = '#000000';
+  public pointFill = '#60B19C';
+  public pointTextFill = '#0000ff';
+  public xAxisTextFill = '#000000';
+  public yAxisTextFill = '#000000';
+  public showXAxis = true;
+  public showYAxis = true;
```
Use a config object of type `QuadrantChartConfig` instead of exposing each config individually.
Then you can use the default values defined in defaultConfig.ts instead of duplicating the values here. 
Also, assignment would be easier.

Instead of
```ts
    quadrantBuilder.quadrantPadding = quadrantChartConfig.quadrantPadding;
    quadrantBuilder.xAxisLabelPadding = quadrantChartConfig.xAxisLabelPadding;
    quadrantBuilder.yAxisLabelPadding = quadrantChartConfig.yAxisLabelPadding;
    quadrantBuilder.xAxisLabelFontSize = quadrantChartConfig.xAxisLabelFontSize;
    quadrantBuilder.yAxisLabelFontSize = quadrantChartConfig.yAxisLabelFontSize;
    quadrantBuilder.quadrantLabelFontSize = quadrantChartConfig.quadrantLabelFontSize;
    quadrantBuilder.quadrantTextTopPadding = quadrantChartConfig.quadrantTextTopPadding;
    quadrantBuilder.pointTextPadding = quadrantChartConfig.pointTextPadding;
    quadrantBuilder.pointLabelFontSize = quadrantChartConfig.pointLabelFontSize;
    quadrantBuilder.pointRadius = quadrantChartConfig.pointRadius;
    quadrantBuilder.xAxisPosition = quadrantChartConfig.xAxisPosition;
    quadrantBuilder.yAxisPosition = quadrantChartConfig.yAxisPosition;
```
Call
```
quadrantBuilder.setConfig(quadrantChartConfig);
```

and inside setConfig 
```ts
public setConfig(updatedConfig: QuadrantChartConfig) {
    this.config = { ...this.config, ...updatedConfig }
}
```

amsubhash replied:
done

sidharthv96 commented on packages/mermaid/src/diagrams/quadrant-chart/quadrantBuilder.ts:
```diff
@@ -0,0 +1,385 @@
+};
+
+export class QuadrantBuilder {
+  private _quadrant1Text = '';
+  private _quadrant2Text = '';
+  private _quadrant3Text = '';
+  private _quadrant4Text = '';
+  private _xAxisLeftText = '';
+  private _xAxisRightText = '';
+  private _yAxisBottomText = '';
+  private _yAxisTopText = '';
+  private _totalHeight = 500;
+  private _totalWidth = 500;
```
Why are these private with getters and setters (which don't do any validation or processing)?
Would be cleaner to make it public no?

amsubhash replied:
I thought of doing the sanitize on the builder but later realized it should be in the db, changed it.

sidharthv96 commented on packages/mermaid/src/diagrams/quadrant-chart/quadrantBuilder.ts:
```diff
@@ -0,0 +1,385 @@
+  public showXAxis = true;
+  public showYAxis = true;
+
+  clear() {
+    this.points = [];
+    this.quadrant1Text = '';
+    this.quadrant2Text = '';
+    this.quadrant3Text = '';
+    this.quadrant4Text = '';
+    this.xAxisLeftText = '';
+    this.xAxisRightText = '';
+    this.yAxisBottomText = '';
+    this.yAxisTopText = '';
+  }
```
You should also reset the config to default values.

sidharthv96 commented on packages/mermaid/src/diagrams/quadrant-chart/quadrantBuilder.ts:
```diff
@@ -0,0 +1,385 @@
+// @ts-ignore: TODO Fix ts errors
+import { scaleLinear } from 'd3';
+
+export type QuadrantPointInputType = { x: number; y: number; text: string };
```
```suggestion
export interface QuadrantPoint { 
  x: number; 
  y: number; 
  text: string 
};
```

amsubhash replied:
done

sidharthv96 commented on packages/mermaid/src/diagrams/quadrant-chart/quadrantBuilder.ts:
```diff
@@ -0,0 +1,385 @@
+export type TextVerticalPos = 'left' | 'center' | 'right';
+export type TextHorizontalPos = 'top' | 'middle' | 'bottom';
+
+export type QuadrantTextType = {
```
```suggestion
export interface QuadrantText {
```

sidharthv96 commented on packages/mermaid/src/diagrams/quadrant-chart/quadrantBuilder.ts:
```diff
@@ -0,0 +1,385 @@
+  rotation: number;
+};
+
+export type QuadrantPointType = {
```
Change all types to interfaces and remove the word `Type` from the names.

sidharthv96 commented on packages/mermaid/src/diagrams/quadrant-chart/quadrantBuilder.ts:
```diff
@@ -0,0 +1,385 @@
+  }
+
+  build(): QuadrantBuildType {
+    const showXAxis = !this.xAxisLeftText && !this.xAxisRightText ? false : this.showXAxis;
+    const showYAxis = !this.yAxisTopText && !this.yAxisBottomText ? false : this.showYAxis;
```
```suggestion
    const showXAxis = this.showXAxis && (this.xAxisLeftText || this.xAxisRightText);
    const showYAxis = this.showYAxis && (this.yAxisTopText || this.yAxisBottomText);
```

amsubhash replied:
It's a nice one, thanks!!, changed.

sidharthv96 commented on packages/mermaid/src/diagrams/quadrant-chart/quadrantBuilder.ts:
```diff
@@ -0,0 +1,385 @@
+        fill: this.quadrant4Fill,
+      },
+    ];
+    quadrants.forEach((quadrant, i) => {
+      // place the text in the center of the box
+      if (this.points.length === 0) {
+        quadrant.text.x = quadrant.x + quadrant.width / 2;
+        quadrant.text.y = quadrant.y + quadrant.height / 2;
+        quadrant.text.horizontalPos = 'middle';
+        // place the text top of the quadrant square
+      } else {
+        quadrant.text.x = quadrant.x + quadrant.width / 2;
+        quadrant.text.y = quadrant.y + this.quadrantTextTopPadding;
+        quadrant.text.horizontalPos = 'top';
+      }
+    });
```
```suggestion
   for(quadrant of quadrants) {
      quadrant.text.x = quadrant.x + quadrant.width / 2;
      // place the text in the center of the box
      if (this.points.length === 0) {
        quadrant.text.y = quadrant.y + quadrant.height / 2;
        quadrant.text.horizontalPos = 'middle';
        // place the text top of the quadrant square
      } else {
        quadrant.text.y = quadrant.y + this.quadrantTextTopPadding;
        quadrant.text.horizontalPos = 'top';
      }
    };
```

sidharthv96 commented on packages/mermaid/src/diagrams/quadrant-chart/quadrantBuilder.ts:
```diff
@@ -0,0 +1,385 @@
+  public pointLabelFontSize = 12;
+  public pointRadius = 5;
+  public points: QuadrantPointInputType[] = [];
+  public xAxisPosition = 'top';
+  public yAxisPosition = 'left';
+  public quadrant1Fill = '#8bc2f3';
+  public quadrant2Fill = '#faebd7';
+  public quadrant3Fill = '#00ffff';
+  public quadrant4Fill = '#f0ffff';
+  public quadrant1TextFill = '#93690e';
+  public quadrant2TextFill = '#8644ff';
+  public quadrant3TextFill = '#e3004d';
+  public quadrant4TextFill = '#000000';
+  public pointFill = '#60B19C';
+  public pointTextFill = '#0000ff';
+  public xAxisTextFill = '#000000';
+  public yAxisTextFill = '#000000';
```
These can be consolidated into a style/theme type variable, like config.

amsubhash replied:
done

sidharthv96 commented on packages/mermaid/src/diagrams/quadrant-chart/quadrantRenderer.ts:
```diff
@@ -0,0 +1,144 @@
+
+  const conf = configApi.getConfig();
+
+  log.debug('Rendering info diagram\n' + txt);
```
wrong log

sidharthv96 commented on packages/mermaid/src/diagrams/quadrant-chart/quadrantRenderer.ts:
```diff
@@ -0,0 +1,144 @@
+  TextVerticalPos,
+} from './quadrantBuilder.js';
+
+export const draw = (txt: string, id: string, _version: string, diagObj: Diagram) => {
```
Does this draw the title if it's present?

amsubhash replied:
I missed it , added it.

sidharthv96 commented:
![image](https://github.com/mermaid-js/mermaid/assets/10703445/94afcdce-04cf-4e62-b694-fff867192375)
The colors don't have enough contrast, we should change the shades slightly. 
Also, can you add an option to draw black lines dividing the quadrants?

sidharthv96 reviewed (changes requested):
This looks wonderful @amsubhash!! 
Few changes to improve the maintainability.

sidharthv96 commented:
<img width="520" alt="image" src="https://github.com/mermaid-js/mermaid/assets/10703445/95646d03-6144-4760-b87e-3baa2a138160">

When I remove the `%%{init`, the axis moves to the top. 
The axis should stay at the bottom if there are points. 

Also, if the labels have an arrow, without end label, the arrow should be rendered. 
Here, x-axis should have arrow, Y shouldn't.
```
quadrantChart
      x-axis "Completeness of Vision ❤" -->
      y-axis Ability to Execute
```

amsubhash commented:
> ![image](https://user-images.githubusercontent.com/10703445/237281920-94afcdce-04cf-4e62-b694-fff867192375.png) The colors don't have enough contrast, we should change the shades slightly. Also, can you add an option to draw black lines dividing the quadrants?

I have used the default theme color here and lighten the primary color,  please check the update chart with border if you still think we need to change the color what I can do is darken the colors if the color is light.

amsubhash commented:
> When I remove the `%%{init`, the axis moves to the top. The axis should stay at the bottom if there are points.
> 
> Also, if the labels have an arrow, without end label, the arrow should be rendered. Here, x-axis should have arrow, Y shouldn't.
> 
> ```
> quadrantChart
>       x-axis "Completeness of Vision ❤" -->
>       y-axis Ability to Execute
> ```
> 
> Also, if there are no points, the axis labels should shift to the center of quadrants' edges.
> 

Fixed all these

amsubhash commented:
> 
> Few more issues I noticed when testing. Some themeVariable is leaking? as the bottom border is green too.
Here I had used the secondaryBorder color now I have changed to primaryBorder color for both
> 
> The `"quadrant1TextFill": "red"` seems to be applied to all labels. 
Fixed
> 
> Can you make sure the arrows on the axis is an SVG or Unicode symbol instead of `-->` ?
Why do we need this? I have not seen this kind of implementation in the codebase. I tried changing to Unicode but it has the same effect.

amsubhash commented on packages/mermaid/src/diagrams/quadrant-chart/quadrantRenderer.ts:
```diff
@@ -0,0 +1,174 @@
+import * as configApi from '../../config.js';
+import { log } from '../../logger.js';
+
+import { configureSvgSize } from '../../setupGraphViewbox.js';
```
done

amsubhash commented:
> ```mermaid
> quadrantChart
>       x-axis Urgent --> Not Urgent 
>       y-axis Not Important --> important
>       quadrant-1 Plan
>       quadrant-2 Do
>       quadrant-3 Deligate
>       quadrant-4 Delete
> ```
> Unable to render rich display
> 
> Diagram error not found. For more information, see https://docs.github.com/get-started/writing-on-github/working-with-advanced-formatting/creating-diagrams#creating-mermaid-diagrams
> 
> quadrantChart
>       x-axis Urgent --> Not Urgent 
>       y-axis Not Important --> important
>       quadrant-1 Plan
>       quadrant-2 Do
>       quadrant-3 Deligate
>       quadrant-4 Delete
> This is failing because, `Not Urgent ` has a space at the end. Should be a fix in jison.

fixed

amsubhash commented on packages/mermaid/src/diagrams/quadrant-chart/parser/quadrant.jison:
```diff
@@ -0,0 +1,188 @@
+ *  https://knsv.github.io/mermaid
+ *  (c) 2015 Knut Sveidqvist
+ *  MIT license.
+ */
```
removed

amsubhash commented on packages/mermaid/src/diagrams/quadrant-chart/quadrantBuilder.ts:
```diff
@@ -0,0 +1,480 @@
+      xAxisPosition: 'top',
+      yAxisPosition: 'left',
+      quadrantInternalBorderStrokeWidth: 2,
+      quadrantExternalBorderStrokeWidth: 3,
```
done

amsubhash commented on packages/mermaid/src/diagrams/quadrant-chart/quadrantBuilder.ts:
```diff
@@ -0,0 +1,480 @@
+
+  getDefaultConfig(): QuadrantBuilderConfig {
+    return {
+      titleText: '',
+      quadrant1Text: '',
+      quadrant2Text: '',
+      quadrant3Text: '',
+      quadrant4Text: '',
+      xAxisLeftText: '',
+      xAxisRightText: '',
+      yAxisBottomText: '',
+      yAxisTopText: '',
+      points: [],
```
done

knsv commented:
> > Few more issues I noticed when testing. Some themeVariable is leaking? as the bottom border is green too.
> 
> Here I had used the secondaryBorder color now I have changed to primaryBorder color for both
> 
> > The `"quadrant1TextFill": "red"` seems to be applied to all labels.
> 
> Fixed
> 
> > Can you make sure the arrows on the axis is an SVG or Unicode symbol instead of `-->` ?
> 
> Why do we need this? I have not seen this kind of implementation in the codebase. I tried changing to Unicode but it has the same effect.

This is just meant as a rendering of the arrow in the final graph without characters, prettier :) Not a showstopper, we can change that to an arrow at a later stage.

Great job!

sidharthv96 reviewed (approved):
Looks good to me! 
Love the unit tests too...

---

The arrows are not a showstopper, just that the `-->` doesn't look "clean". ⭢ ⭡Does using these unicode help?

tobie commented:
This is amazing! thank you so much for doing the work!

amsubhash commented:
> <img alt="image" width="283" src="https://user-images.githubusercontent.com/10703445/238281316-e3e41d42-daef-4bea-822e-1c87bfeea9cf.png"> <img alt="image" width="271" src="https://user-images.githubusercontent.com/10703445/238280942-3a38bd46-07cd-45e0-b5a5-75615f94b173.png">
> Unicode does seem to work.
> 
> <img alt="image" width="931" src="https://user-images.githubusercontent.com/10703445/238281040-23f760f8-4266-40f8-bd7f-3162c279a58c.png">

Ok I got it I was thinking of changing only the `>`  as it can be a special case of html rendering issue. any way I can add it later.

Overall it was a nice experience I had fun building the feature, thanks @sidharthv96  for the detailed review, I learned a lot while developing it specially, `jison` and `cypress`.

padorob commented:
Do you think it would be possible to have a third dimension for the relative bubble size?
So that one could create also things like:
![immagine](https://github.com/mermaid-js/mermaid/assets/125264250/56454329-39e4-485a-bc33-6b07571cb250)
